clients/drcachesim/scheduler/noise_generator.h

```diff
@@ -0,0 +1,132 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _NOISE_GENERATOR_H_
+#define _NOISE_GENERATOR_H_ 1
+
+#include "reader.h"
+#include "scheduler.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
+ * Generates synthetic #dynamorio::drmemtrace::memref_t trace entries in a single thread
+ * and presents them via an iterator interface to the scheduler.
+ */
+class noise_generator_t : public reader_t {
+public:
+    noise_generator_t();
+
+    noise_generator_t(addr_t pid, addr_t tid, uint64_t num_records_to_generate);
+
+    virtual ~noise_generator_t();
+
+    bool
+    init() override;
+
+    std::string
+    get_stream_name() const override;
+
+protected:
+    // Makes sure the noise records generated by generate_trace_entry() are preceded by
+    // TRACE_TYPE_THREAD, TRACE_TYPE_PID, TRACE_MARKER_TYPE_TIMESTAMP and followed by
+    // TRACE_TYPE_THREAD_EXIT.
+    virtual trace_entry_t *
+    read_next_entry() override;
+
+    // Contains the main logic to generate noise records.
+    virtual trace_entry_t
+    generate_trace_entry();
+
+    // This counter does not count TRACE_TYPE_THREAD, TRACE_TYPE_PID, and
+    // TRACE_MARKER_TYPE_TIMESTAMP. The idea is that when the user wants to generate at
+    // at least one record, tid, pid, and timestamp always have to be there, otherwise
+    // the scheduler will report an error.
+    uint64_t num_records_to_generate_ = 0;
+    addr_t pid_ = 0;
+    addr_t tid_ = 0;
+
+private:
+    trace_entry_t entry_ = {};
+    bool pid_generated_ = false;
+    bool tid_generated_ = false;
+    bool marker_timestamp_generated_ = false;
+};
+
+/**
+ * Contains metadata information to drive the noise generation.
+ */
+struct noise_generator_info_t {
+    // TODO i#7216: temporary default values.
+    uint64_t num_processes = 1;
+    uint64_t num_threads_per_process = 1;
+    uint64_t num_records_to_generate = 1000;
+};
+
+/**
+ * Factory to create noise_generator_t.
+ */
+template <typename RecordType, typename ReaderType> class noise_generator_factory_t {
+public:
+    typedef scheduler_tmpl_t<RecordType, ReaderType> sched_type_t;
+
+    void
+    init();
+
+    bool
+    is_enabled();
```

Comment:
This "enabled" seems odd: it's not used to gate the add_ function; it is in fact ignored by the internal code of this class.  It doesn't seem like it belongs inside the class.

Reply:
Right, moved to the analyzer instead.